DNS support for static nodes and discovery

prd-fox opened the pull request:
Adds full DNS support for Quorum.

This includes:

use of DNS hostnames in the static-nodes.json file, so that we can connect to our static nodes even when the IP address of the DNS name has changed, meaning less updating of the static node definition file. This is an improvement to previous behaviour which resolved the DNS hostname once at startup and then only tracked the IP address

prd-fox committed:
Update raft to accept a hostname instead of only an IP address
Add a switch to enable DNS changes in raft to ensure default behaviour is backwards-compatible with older versions.
Add ability to deserialise old raft address information
Remove IP and undecoded elements from output via raft API.

Allow P2P layer to use the FQDN

Add hostname flag to specify the DNS hostname to use during discovery for this node
Add hostname of this node to ping packets, so that the receiver adds the DNS name to their node bucket

Add Quorum specific UDP classes and types, which are a modified version of discv4

Make bootnode listen for Quorum Discovery messages

Remove double-decode of rlp elements by creating the RLP stream outside.
Add forward-compatibility by adding a tail of undecoded RLP elements that are ignored.

Add document for command line arguments, link to upstream arguments and
including DNS specific changes.

prd-fox committed:
Export struct members so they can be RLP decoded.

prd-fox committed:
Fix doc typo and make doc use clearer language.

prd-fox committed:
Add hostname parameter to TOML file and run goimports on badly formatted
files.

prd-fox committed:
Review feedback:
move docs to correct place and update doc index
Fix mehod comments and Quorum location comments

prd-fox committed:
Review feedback:
Move new "hostname" flag to be under Quorum options instead of
networking options.

prd-fox committed:
Review feedback:
Take upstream change for hostname parsing found at
https://github.com/ethereum/go-ethereum/pull/18524

prd-fox committed:
Review feedback:
Keep raft CLI arg name in line with other raft args by keeping as all
one word

prd-fox committed:
Fallback to IP address if no hostname specified for discovered nodes.

prd-fox committed:
Add document describing DNS in more detail.
The node will not start if the provided hostname flag does not resolve
to an IP address, to prevent a user accidentally mistyping their
intended address.

prd-fox committed:
Remove debug line and expand document explaining the features of DNS
support.

prd-fox committed:
Rearrange document to make it clearer why DNS is always enabled for
IBFT and Clique

prd-fox committed:
Include DNS document in root tree so it appears on the docs site.

Krish1979 committed:
Merge branch 'master' into dns-support

trung commented on mkdocs.yml:
```diff
@@ -18,6 +18,7 @@ nav:
         - Creating a Network From Scratch: Getting Started/Creating-A-Network-From-Scratch.md
         - Running Quorum: Getting Started/running.md
         - Quorum API: Getting Started/api.md
+        - CLI arguments: Getting Started/command-line-arguments.md
```
I think we don't really need this page as it seems to be a duplicated of `geth --help`

prd-fox replied:
I can expand it a bit more as I agree that they are quite similar. Ideally `geth --help` should be a one liner of the flag, whereas this page can give a bit more description.
The flags will likely be explained further (or repeated) in their respective doc pages, but it is helpful IMO to have an aggregate page of all flags.

trung replied:
CLI flags are expected to be explained well in their respected features doc. Creating a central page  is useful but I see some maintenance overhead and duplication of docs here.

prd-fox replied:
We can add it later with a more comprehensive set of options.

trung commented on mkdocs.yml:
```diff
@@ -75,7 +76,9 @@ nav:
         - Getting Started: Cakeshop/Getting started.md
         - Cakeshop FAQ: Cakeshop/Cakeshop FAQ.md
-    - Product Roadmap: roadmap.md         
+    - Quorum Features:
```
As all features are in the 1st level, DNS is deserved in the first level too.
Otherwise we could move others to `Enterprise Features` menu. But i prefer the former.

prd-fox replied:
It was intended to be a place to put features that don't fit the "main" categories. Adding them to the top level would create too much clutter, but happy to consider alternatives.

trung commented on p2p/shared_udp_connection.go:
```diff
@@ -0,0 +1,84 @@
+// UDPReader implements a shared connection. Write sends messages to the
+// underlying connection while read returns messages that were found
+// unprocessable and sent to the unhandled channel by the primary listener.
+type UDPReader interface {
```
This interface has number of methods that deal with UDP including read/write/close. It's deserved a more descriptive name than just a `Reader`.  And it seems a dup of local interface `conn` from `p2p/discover` package.

Is there a chance to move some to `discover` package or expose from `discover` package instead?

prd-fox replied:
Agree about the name.
Do you think I should export the `conn` interface from `discover` instead?

trung commented on p2p/discover/udp_quorum.go:
```diff
@@ -0,0 +1,479 @@
+		Expiration uint64
+		// Ignore additional fields (for forward compatibility).
+
+		IsQuorum   bool
```
Why do we need this field here?

Same for for `neighborsQuorum`

prd-fox replied:
I considered this a cheap way to differentiate the packet from a DiscV4 packet. A DiscV4 packet will fail decoding when checked in the Quorum DiscV4 decoder, and get passed on appropriately.

trung commented on p2p/discover/udp_quorum.go:
```diff
@@ -0,0 +1,479 @@
+	// Neighbors replies are sent across multiple packets to
+	// stay below the 1280 byte limit. We compute the maximum number
+	// of entries by stuffing a packet until it grows too large.
+	maxNeighborsQuorum = 3
```
this value is calculated in `udp.go#init()`, why do we hard code here?

prd-fox replied:
The value calculated in `udp.go#init()` (12, if I remember correctly) is so because a record in discovery V4 is a fixed size, so you can always fit 12 records in a single packet. 
Since hostnames are a variable size, this is no longer true. 3 was fairly arbitrary but unlikely to ever reach the limit.
I'll change it to check the size after each node has been added and see if it has gone over the limit.

trung reviewed (commented):
Initial quick feedbacks

libby commented on p2p/enode/urlv4.go:
```diff
@@ -103,6 +103,37 @@ func NewV4(pubkey *ecdsa.PublicKey, ip net.IP, tcp, udp, raftPort int) *Node {
 	return n
 }
 
+// Quorum
```
We could break out the shared logic between `NewV4` and `NewV4Hostname` into a private function `newV4` this would make it easier to see there is no change in the logic except for hostname vs IP type. 
```
// NewV4 creates a node from discovery v4 node information. The record
// contained in the node has a zero-length signature.
func NewV4(pubkey *ecdsa.PublicKey, ip net.IP, tcp, udp, raftPort int) *Node {
	var r enr.Record
	if ip != nil {
		r.Set(enr.IP(ip))
	}
        if raftPort != 0 { // Quorum specific (TODO: can we remove this here? and from the function sig?)
		r.Set(enr.RaftPort(raftPort))
	}
	return newV4(pubkey, r, tcp, udp, raftPort)
}
// broken out from `func NewV4` (above) same in upstream go-ethereum, but taken out
// to avoid code duplication b/t NewV4 and NewV4Hostname
func newV4(pubkey *ecdsa.PublicKey, r enr.Record, tcp, udp, raftPort int) *Node {
     if udp != 0 {
       r.Set(enr.UDP(udp))
     }
     if tcp != 0 {
       r.Set(enr.TCP(tcp))
     }

     if raftPort != 0 { // Quorum
       r.Set(enr.RaftPort(raftPort))
     }

     signV4Compat(&r, pubkey)
     n, err := New(v4CompatID{}, &r)
     if err != nil {
       panic(err)
     }
     return n
}
// Quorum

// NewV4Hostname creates a node from discovery v4 node information. The record
// contained in the node has a zero-length signature. It sets the hostname of
// the node instead of the IP address.
func NewV4Hostname(pubkey *ecdsa.PublicKey, hostname string, tcp, udp, raftPort int) *Node {
	var r enr.Record
	if hostname != "" {
		r.Set(enr.Hostname(hostname))
	}
	if raftPort != 0 {
		r.Set(enr.RaftPort(raftPort))
	}
	return newV4(pubkey, r, tcp, udp, raftPort)
}

// End-Quorum
```
Additionally, it seems the `raftPort` is only set in `parseComplete` so the signature for `NewV4` could lose the raftPort `func NewV4(pubkey *ecdsa.PublicKey, ip net.IP, tcp, udp int) *Node` and thus match upstream (https://github.com/ethereum/go-ethereum/blob/master/p2p/enode/urlv4.go#L85), and the Quorum specific `raftport` could be set only in the Quorum specific `NewV4Hostname(pubkey *ecdsa.PublicKey, hostname string, tcp, udp, raftPort int)` before calling `func newV4(pubkey *ecdsa.PublicKey, r enr.Record, tcp, udp int) *Node ` (TODO: check me on this).

prd-fox replied:
I made this change, although I originally kept them separate to avoid changing upstream code (only adding onto). As far as the raft port is concerned, you may be right, but I don't plan on messing around with that as it doesn't pertain to this change.

libby commented on p2p/enode/urlv4.go:
```diff
@@ -124,19 +155,17 @@ func parseComplete(rawurl string) (*Node, error) {
 	// Ensure the IP is 4 bytes long for IPv4 addresses.
 	if ipv4 := ip.To4(); ipv4 != nil {
 		ip = ipv4
 	}
```
This logic has changed slight in upstream https://github.com/ethereum/go-ethereum/blob/master/p2p/enode/urlv4.go#L130 we might want to look into syncing it up in this PR. Additionally, a quorum specific comment might be helpful. Something like: 
```
 // Quorum: is only testing the hostname resolves to  a validate the IPv4 addr, but unlike upstream geth, it does not pass the resolved IP on, but rather uses the hostname.
// Parse the IP address.
...
 ```

prd-fox replied:
I'm not enthusiastic on pulling in individual upstream changes unless its to fix a critical bug. We will get this change when we upgrade, and we can keep our version (1.8.18) as close to the same version upstream.
As you note in your comment, we don't use the IP anyway, rather passing in the hostname field (which may be an IP).
Let me know if you feel strongly that we do pull this change in.

libby replied:
I don't feel strongly, it was only because the change from upstream was being brought in and 
 that had fallen behind (https://github.com/ethereum/go-ethereum/commit/b90cdbaa79cfe438aab0f1389d35980f3d38ec84#diff-0532cc1dc91dc5633d79c93dac1cd87bR128). Removing the logic (as you have now done) seems fine to me and less confusing as it isn't being passed in.

libby commented on p2p/enode/urlv4_test.go:
```diff
@@ -55,39 +54,39 @@ var parseNodeTests = []struct {
 			52150,
 			52150,
 			0,
 		),
```
Should we keep the `NewV4` test and just add `NewV4Hostname` tests? Could even be in another file like `p2p/enode/urlv4_quorum_test.go`.

prd-fox replied:
That sounds like a good idea.

prd-fox replied:
Hm, then again, since the tests make use of `parseComplete` which has been modified to use `NewV4Hostname`, the original tests don't make sense. We could duplicate the entire functionality of `parseComplete` and all its callsites, but that seems like a lot of effort for little gain.

libby reviewed (changes requested):
Initial feedback =)

prd-fox committed:
Move out common functionality between enode object creation (IP vs
Hostname).

prd-fox committed:
Remove commandline page since it only contains changes related to DNS.
It should be included later with a comprehensive listing of Quorum
specific options.

prd-fox committed:
Merge branch 'dns-support' of https://github.com/prd-fox/quorum into dns-support

trung committed:
p2p: use the same interface to avoid duplication

trung commented on p2p/server.go:
```diff
@@ -533,25 +520,50 @@ func (srv *Server) setupDiscovery() error {
+			tab = tabCasted
+		}
+
+		_, err := discover.ListenUDPExisting(quorumSharedConnection, srv.localnode, cfg, tab)
```
Are these lines from 551 to 566 redundant? Lines 547-550 can be moved to 529?

jbhurat commented on p2p/discover/udp.go:
```diff
@@ -268,6 +268,39 @@ func newUDP(c conn, ln *enode.LocalNode, cfg Config) (*Table, *udp, error) {
+// Quorum
+
+// ListenUDPExisting uses an existing table that listens for UDP packets on the connection.
+func ListenUDPExisting(c conn, ln *enode.LocalNode, cfg Config, tab *Table) (*Table, error) {
```
Existing `ListenUDP()` and `newUDP()` functions are no longer referenced anywhere, is the plan to delete those?

prd-fox replied:
I hadn't planned to, in order to keep the diff from upstream smaller. `newUDP()` is still used in the upstream tests (`udp_test.go`).

trung commented on raft/handler.go:
```diff
@@ -656,11 +675,24 @@ func (pm *ProtocolManager) addPeer(address *Address) {
+		// attempt to look up IP addresses if host is a FQDN
+		lookupIPs, err := net.LookupIP(string(address.Hostname))
+		if err != nil {
+			panic(err.Error())
```
Node is not able to start. This happens when peers' hostnames and IPs have not yet registered in DNS. 

It is a challenge for environments where  DNS entry is only available when VM starts up.

jbhurat reviewed (commented):
I see that there is a lot of duplication between `udp.go` and `udp_quorum.go` is it possible to re-evaluate `udp.go` and `udp_quorum.go` to see there is no code duplication

prd-fox committed:
Remove discovery based DNS support.

prd-fox committed:
Fix newline formatting and update documentation for the removal of
discovery based DNS support.

prd-fox committed:
Merge branch 'master' into dns-support

prd-fox committed:
Update node parsing tests to include hostnames.

prd-fox committed:
Merge remote-tracking branch 'prd-fox/dns-support' into dns-support

prd-fox committed:
Removed hostname from discovery test.

prd-fox committed:
Allow continuous hostname resolution for static nodes, not just at boot
time. This allows us to continuously try to connect to static nodes
without restarting.

prd-fox committed:
Moved a test to not be included in the full suite. This is because the
hostname does not resolve to a valid IP, so the output will not match
the input as the current test setup expects. The output will be caught
in other part of the systems and handled appropriately.

prd-fox committed:
Update documentation to reflect the fact that defined bootnodes will
keep resolving the hostname instead of just at startup.

Ran goimports.

trung reviewed (dismissed):
LGTM

jbhurat reviewed (dismissed):
LGTM

prd-fox committed:
Make snapshot struct names clearer as to their contents, instead of "old"
vs "new", which will not having meaning in the future.

prd-fox committed:
Add trace log for when hostname resolution fails, letting the user know
how the resolution failed

jpmsam committed:
Merge branch 'master' into dns-support
